pkg/controller/highperf/highperf_controller.go

```diff
@@ -0,0 +1,242 @@
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting high perf controller")
+	defer klog.Info("Shutting down high perf controller")
+	controllerManagerMetrics.ControllerStarted("highperf")
+	defer controllerManagerMetrics.ControllerStopped("highperf")
+
+	if !cache.WaitForNamedCacheSync("node", stopCh, c.nodeInformer.Informer().HasSynced) {
+		return
+	}
+
+	if !cache.WaitForNamedCacheSync("networks", stopCh, c.nwInformer.HasSynced) {
+		return
+	}
+
+	if !cache.WaitForNamedCacheSync("gkenetworkparamset", stopCh, c.gnpInformer.HasSynced) {
+		return
+	}
+
+	for i := 0; i < numWorkers; i++ {
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) runWorker(ctx context.Context) {
+	for c.processNextItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	err := c.processNode(ctx, key.(string))
+	c.handleErr(err,key)
+	return true
+}
+
+func (c *Controller)handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	// This controller retries 5 times if something goes wrong. After that, it stops trying.
+	if c.queue.NumRequeues(key) < 5 {
+		klog.Warningf("Error while updating GKENetworkParamSet object, retrying %v: %v", key, err)
+
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	c.queue.Forget(key)
+	// Report to an external entity that, even after several retries, we could not successfully process this key
+	utilruntime.HandleError(err)
+	klog.Errorf("dropping node  %q out of the queue: %v", key, err)
+}
+
+func (c *Controller) processNode(ctx context.Context, key string) error {
+	node, err := c.nodeInformer.Lister().Get(key)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("fetching object with key %s from store failed with %v", key, err)
+	}
+
+	//northInterface and node-status annotations need to be ready before we move on
+	if node.Annotations[networkv1.NorthInterfacesAnnotationKey] == "" || node.Annotations[networkv1.NodeNetworkAnnotationKey] == "" {
+		return nil
+	}
+
+	// check if northInterface and network-status have entries for same network
+	northInterfaces, err := networkv1.ParseNorthInterfacesAnnotation(node.Annotations[networkv1.NorthInterfacesAnnotationKey])
+	if err != nil {
+		return fmt.Errorf("parsing northInterface annotation: %v", err)
+	}
+	nodeNetworks, err := networkv1.ParseNodeNetworkAnnotation(node.Annotations[networkv1.NodeNetworkAnnotationKey])
+	if err != nil {
+		fmt.Errorf("parsing network-status annotation: %v", err)
+	}
+	var readyNetworks networkv1.NorthInterfacesAnnotation
+
+	// create a list of networks present in both the northInterface and network-status annotations
+	statusMap := make(map[string]string)
```

Comment:
nit: a more canonical way is to use map[string]struct{}, and when putting things into the "set" do statusMap[key] = struct{}{}